File: AppCheckCore/Tests/Unit/Core/GACAppCheckTokenResultTests.m
/*
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import <XCTest/XCTest.h>

#import "AppCheckCoreProvider/Sources/Public/AppCheckCoreProvider/GACAppCheckToken.h"
#import "AppCheckCoreProvider/Sources/Public/AppCheckCoreProvider/GACAppCheckTokenResult.h"

static NSString *const kTestTokenValue = @"test-token";
/// Placeholder value that indicates failure: `{"error":"UNKNOWN_ERROR"}` encoded as base64
static NSString *const kPlaceholderTokenValue = @"eyJlcnJvciI6IlVOS05PV05fRVJST1IifQ==";
static NSString *const kTestErrorDomain = @"TestErrorDomain";
static NSInteger const kTestErrorCode = 42;

@interface GACAppCheckTokenResult (Tests)

+ (GACAppCheckToken *)placeholderToken;

@end

@interface GACAppCheckTokenResultTests : XCTestCase
@end

@implementation GACAppCheckTokenResultTests

- (void)testInitWithToken {
  NSDate *expectedExpirationDate = [NSDate dateWithTimeIntervalSince1970:1693314000.0];
  NSDate *expectedReceivedAtDate = [NSDate dateWithTimeIntervalSince1970:1693317600.0];
  GACAppCheckToken *expectedToken = [[GACAppCheckToken alloc] initWithToken:kTestTokenValue
                                                             expirationDate:expectedExpirationDate
                                                             receivedAtDate:expectedReceivedAtDate];

  GACAppCheckTokenResult *tokenResult =
      [[GACAppCheckTokenResult alloc] initWithToken:expectedToken];

  XCTAssertEqualObjects(tokenResult.token, expectedToken);
  XCTAssertNil(tokenResult.error);
}

- (void)testInitWithError {
  NSError *expectedError = [NSError errorWithDomain:kTestErrorDomain
                                               code:kTestErrorCode
                                           userInfo:nil];

  GACAppCheckTokenResult *tokenResult =
      [[GACAppCheckTokenResult alloc] initWithError:expectedError];

  XCTAssertEqualObjects(tokenResult.token.token, kPlaceholderTokenValue);
  XCTAssertNotNil(tokenResult.error);
  XCTAssertEqualObjects(tokenResult.error, expectedError);
}

- (void)testInitWithTokenAndError {
  GACAppCheckToken *placeholderToken = [GACAppCheckTokenResult placeholderToken];
  NSError *expectedError = [NSError errorWithDomain:kTestErrorDomain
                                               code:kTestErrorCode
                                           userInfo:nil];

  GACAppCheckTokenResult *tokenResult =
      [[GACAppCheckTokenResult alloc] initWithToken:placeholderToken error:expectedError];

  XCTAssertEqualObjects(tokenResult.token, placeholderToken);
  XCTAssertNotNil(tokenResult.error);
  XCTAssertEqualObjects(tokenResult.error, expectedError);
}

- (void)testPlaceholderToken {
  NSDate *expectedExpirationDate = [NSDate distantPast];
  NSDate *expectedReceivedAtDate = [NSDate date];  // Current time

  GACAppCheckToken *placeholderToken = [GACAppCheckTokenResult placeholderToken];

  XCTAssertEqualObjects(placeholderToken.token, kPlaceholderTokenValue);
  // Verify that the placeholder token's received at time is approximately equal to current time.
  XCTAssertEqualWithAccuracy(
      [placeholderToken.receivedAtDate timeIntervalSinceDate:expectedReceivedAtDate], 0, 5.0);
  XCTAssertEqualObjects(placeholderToken.expirationDate, expectedExpirationDate);
}

@end

File: AppCheckCore/Tests/Unit/Core/GACAppCheckTokenResultTests.swift
import XCTest
@testable import AppCheckCore

private let kTestTokenValue = "test-token"
/// Placeholder value that indicates failure: `{"error":"UNKNOWN_ERROR"}` encoded as base64
private let kPlaceholderTokenValue = "eyJlcnJvciI6IlVOS05PV05fRVJST1IifQ=="
private let kTestErrorDomain = "TestErrorDomain"
private let kTestErrorCode = 42

class AppCheckCoreTokenResultTests: XCTestCase {
  
  func testInitWithToken() {
    let expectedExpirationDate = Date(timeIntervalSince1970: 1693314000.0)
    let expectedReceivedAtDate = Date(timeIntervalSince1970: 1693317600.0)
    let expectedToken = AppCheckCoreToken(token: kTestTokenValue,
                                         expirationDate: expectedExpirationDate,
                                         receivedAtDate: expectedReceivedAtDate)
    
    let tokenResult = AppCheckCoreTokenResult(token: expectedToken)
    
    XCTAssertEqual(tokenResult.token, expectedToken)
    XCTAssertNil(tokenResult.error)
  }
  
  func testInitWithError() {
    let expectedError = NSError(domain: kTestErrorDomain,
                                code: kTestErrorCode,
                                userInfo: nil)
    
    let tokenResult = AppCheckCoreTokenResult(error: expectedError)
    
    XCTAssertEqual(tokenResult.token.token, kPlaceholderTokenValue)
    XCTAssertNotNil(tokenResult.error)
    XCTAssertEqual(tokenResult.error as NSError?, expectedError)
  }
  
  func testInitWithTokenAndError() {
    let placeholderToken = AppCheckCoreTokenResult.placeholderToken()
    let expectedError = NSError(domain: kTestErrorDomain,
                                code: kTestErrorCode,
                                userInfo: nil)
    
    let tokenResult = AppCheckCoreTokenResult(token: placeholderToken, error: expectedError)
    
    XCTAssertEqual(tokenResult.token, placeholderToken)
    XCTAssertNotNil(tokenResult.error)
    XCTAssertEqual(tokenResult.error as NSError?, expectedError)
  }
  
  func testPlaceholderToken() {
    let expectedExpirationDate = Date.distantPast
    let expectedReceivedAtDate = Date() // Current time
    
    let placeholderToken = AppCheckCoreTokenResult.placeholderToken()
    
    XCTAssertEqual(placeholderToken.token, kPlaceholderTokenValue)
    // Verify that the placeholder token's received at time is approximately equal to current time.
    XCTAssertEqual(placeholderToken.receivedAtDate.timeIntervalSince(expectedReceivedAtDate), 0, accuracy: 5.0)
    XCTAssertEqual(placeholderToken.expirationDate, expectedExpirationDate)
  }
}
